<hr />

<p>title: ""
author: "Mark Dunning"
output:
  html<em>notebook:
    toc: yes
    toc</em>float: yes
  html<em>document:
    df</em>print: paged
    toc: yes
editor_options:</p>

<h2>  chunk<em>output</em>type: inline</h2>

<h1>Hands-on NGS Analysis in Galaxy</h1>

<h3>Sheffield Bioinformatics Core</h3>

<p><img src="media/logo-sm.png" align=right></p>

<p>web : <a href="http://sbc.shef.ac.uk">sbc.shef.ac.uk</a> <br />
twitter: <a href="https://twitter.com/SheffBioinfCore">SheffBioinfCore</a> <br />
email: <a href="bioinformatics-core@sheffield.ac.uk">bioinformatics-core@sheffield.ac.uk</a></p>

<hr />

<h1>Tutorial Overview</h1>

<p>This tutorial will cover the basics of NGS analysis using Galaxy; a open-source web-based platform for the analysis of biological data. You should gain an appreciation of the tasks involved in a typical NGS analysis and be comfortable with the outputs generated by a sequencing service. </p>

<p>Parts of this tutorial are based on the NGS tutorial from the Galaxy Project</p>

<ul>
<li><a href="https://galaxyproject.org/tutorials/ngs/">Galaxy NGS Tutorial</a></li>
</ul>

<p>The tutorial covers the following steps in our analysis pipeline</p>

<p><img src="media/pipeline.png" alt="" title="" /></p>

<hr />

<h2>Background</h2>

<h4>Where do the data in this tutorial come from?</h4>

<p>The data for this tutorial are publicaly-available exome sequencing data <em>downsampled</em> to the BRCA2 region from a fictitious patient. We will use these data throughout the course to call variants, filter and discuss the clinical impact of any mutations. </p>

<h1>Section 1: Preparation</h1>

<h4>1.  Register as a new user on one of the public Galaxy servers</h4>

<ul>
<li>https://usegalaxy.org/</li>
<li>https://usegalaxy.org.au</li>
<li>https://usegalaxy.eu</li>
</ul>

<p><strong>Make sure you check your email to activate your account</strong></p>

<h4>2.  Import the data for the workshop.</h4>

<p>We can going to import the <a href="https://en.wikipedia.org/wiki/FASTQ_format"><em>fastq</em> files</a> for this experiment. This is a standard format for storing raw sequencing reads and their associated quality scores. However, as we will see, the representation of the quality scores has changed over time.</p>

<p>You can import the data by:</p>

<ol>
<li>In the tool panel located on the left, under Basic Tools select <strong>Get
Data > Upload File</strong>. Click on the <strong>Paste/Fetch data</strong> button on the
bottom section of the pop-up window.</li>
<li><p>Upload the sequence data by selecting the files <code>JoeBlogsBRCAPanel_R1.fastq</code> and <code>JoeBlogsBRCAPanel_R2.fastq</code>. Make sure the type is specified as '<strong>fastqsanger</strong>'
when uploading.</p></li>
<li><p>You should now have these 2 files in your history:</p>

<ul>
<li><code>JoeBlogsBRCAPanel_R1.fastq</code></li>
<li><code>JoeBlogsBRCAPanel_R2.fastq</code></li>
</ul></li>
</ol>

<h1>Section 2: Fastq file format</h1>

<p>You can view the files you just uploaded by clicking the <strong>eye icon</strong> the history item. The first few lines should read as follows</p>

<p><strong>JoeBlogsBRCAPanel_R1.fastq</strong></p>

<p><code>
@HWI-D00461:188:HVGY2BCXY:1:1101:1363:84148/1
TGTGTCATTTCTATTATCTTTGGAACAACCATGAATTAGTCCCTTGGGGTTTTCAAATGCTGCACACTGACTCACACATTTATTTGGTTCTGTTTTTGCCTTCCCTNN
+
DDDDDIHIIIIIHIIIIIIIIIIHIIIIIIIIIHIIIIIIHHIIIIIIIHIIIHIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIHHIDHHIHC#
</code></p>

<p><strong>JoeBlogsBRCAPanel_R2.fastq</strong></p>

<p><code>
@HWI-D00461:188:HVGY2BCXY:1:1101:1363:84148/2
TGGAAAGACTTTTGGGGGGGGGAGTATTTTTCTTGTTTCTGGTTTTGGTTTTTTTGATCCGGGAAAGATTTTGTTTTTTGGAGGTTGGACTTTTGGGGAGGGGAAAAN
+
&lt;00&lt;&lt;1111&lt;1&lt;11/0/&lt;///&lt;/&lt;0111DF11&lt;&lt;11111&lt;11111&lt;1/1&lt;1D1///&lt;&lt;11&lt;&lt;&lt;/&lt;1111&lt;&lt;11&lt;DD1&lt;//&lt;110&lt;11/11&lt;0&lt;0&lt;/0&lt;0&lt;&lt;-//&lt;&lt;FE
</code></p>

<p>Each read is described by 4 lines in the file:-</p>

<p><img src="media/fastq-header.png"></p>

<p>The first line is the unique identifier for each sequenced read. It can be used to encode information such as the <em>sequencing machine</em>, <em>flow cell</em> and <em>lane</em> that the read was generated from and the physical coordinates on the lane.  </p>

<p>The quality scores are <a href="http://ascii-code.com/">ASCII</a> representations of how confident we are that a particular base has been called correctly. Letters that are further along the alphabet indicate higher confidence. This is important when trying to identify types of genome variation such as single base changes, but is also indicative of the overall quality of the sequencing. Different scales have been employed over time (resulting in a different set of characters appearing in the file). We will need to tell Galaxy which scale has been used in order that we can process the data correctly; hence why we explictly stated the files were of type <strong>fastqsanger</strong> when we uploaded them. </p>

<h3>Deriving the Quality Score</h3>

<p>First of all, we convert the base-calling probability (p) into a <code>Q</code> score using the formula</p>

<ul>
<li>Quality scores $$ Q = -10log_{10}p$$
<ul>
<li>Q = 30, p=0.001</li>
<li>Q = 20, p=0.01</li>
<li>Q = 10, p=0.1</li>
</ul></li>
<li>These numeric quanties are <em>encoded</em> as <a href="http://ascii-code.com/"><strong>ASCII</strong></a> code
<ul>
<li>At least 33 to get to meaningful characters
(https://en.wikipedia.org/wiki/FASTQ_format)</li>
</ul></li>
</ul>

<p><img src="media/phred.png" alt="" title="" />      </p>

<h3>Quality Scores to probabilities</h3>

<ul>
<li>look-up the ASCII code for each character</li>
<li>subtract the offset to get the Q score</li>
<li>convert to a probability using the formula:-</li>
</ul>

<p>$$ p = 10^{-Q/10} $$</p>

<p>Let's see this calculation for the first few bases of the first read in <code>JoeBlogsBRCAPanel_R1.fastq</code>; <code>DDDDDIHI....</code></p>

<p>Character  | Code | Minus 33 Offset | Probability
------------- | ------------- | ------------- | -------------
D  | 68 | 35 | 0.0003162278
D  | 68 | 35 | 0.0003162278
D  | 68 | 35 | 0.0003162278
D  | 68 | 35 | 0.0003162278
D  | 68 | 35 | 0.0003162278
I  | 73 | 40 | 0.0001000000
H  | 72 | 39 | 0.0001258925
I  | 73 | 40 | 0.0001000000</p>

<p>In practice, we don't have to convert the values as we have software that will do this automatically</p>

<hr />

<h1>Section 3: Quality assessment with FastQC</h1>

<p><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> is a popular tool from <a href="https://www.bioinformatics.babraham.ac.uk/index.html">Babraham Institute Bioinformatics Group</a> used for <em>quality assessment</em> of sequencing data. Most Bioinformatics pipelines will use FastQC, or similar tools in the first stage of the analysis. The <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/">documentation</a> for FastQC will help you to interpret the plots and stats produced by the tool. A traffic light system is used to alert the user's attention to possible issues. </p>

<ul>
<li>From the left hand tool panel in Galaxy, under <em>NGS ANALYSIS</em>, select <em>NGS: QC and manipulation -> FastQC</em></li>
<li>Select one of the FASTQ files as input and <em>Execute</em> the tool.</li>
<li>When the tool finishes running, you should have an HTML file in your History. Click on the eye icon to view the various quality metrics.</li>
</ul>

<p>The most important image is whether the base quality decreases significantly over the length of the read</p>

<p><img src="media/fastqc-2a.png" alt="" title="" /></p>

<p>Good quality data should look something like:-</p>

<p><img src="media/fastqc-2b.png" alt="" title="" /></p>

<p>Look at the generated FastQC metrics for your uploaded fastq files. This data looks pretty good - high per-base quality scores (most above 30).</p>

<p>All is not lost if we observe poor quality bases towards the end of the read. There are a number of <em>trimming</em> options that we can use for NGS data and some of these are availble through Galaxy. Check out the <a href="https://galaxyproject.org/tutorials/ngs/#trimming-reads">Trimmming Reads</a> section of the Galaxy NGS tutorial if you are interested in how we can trim our reads.</p>

<p>It is also worth bearing in mind that the tool is blind to the particular type of sequencing you are performing (i.e. whole-genome, ChIP-seq, RNA-seq) and the organism being sequenced, so some warnings might be expected due to the nature of your experiment. For instance, there are known sequencing composition biases that can occur at the beginning of RNA-seq reads. </p>

<h1>Section 4: Alignment</h1>

<p>We don't really spend much time look at <em>fastq</em> files, as most of our time is spent with <em>aligned</em> reads. i.e. we have used some software to tell us whereabouts in the genome each read belongs to. This will <em>usually</em> be performed for you as part of a sequencing service, but it is good to get an appreciation of the steps involved.</p>

<p>In this section we map the reads in our FASTQ files to a reference genome. </p>

<p>A plethora of different tools have been written to perform this task, and we will not describe it in detail. Links to some key publications are given below:-</p>

<ul>
<li>2009 Bowtie 1 - <a href="http://genomebiology.com/content/10/3/R25">Langmead et al</a></li>
<li>2012 Bowtie 2 - <a href="http://www.nature.com/nmeth/journal/v9/n4/full/nmeth.1923.htm">Langmead and Salzberg</a></li>
<li>2009 BWA - <a href="http://bioinformatics.oxfordjournals.org/content/25/14/1754.long">Li and Durbin</a></li>
<li>2010 BWA - <a href="http://bioinformatics.oxfordjournals.org/content/26/5/589">Li and Durbin</a></li>
<li>2013 BWA-MEM - <a href="http://arxiv.org/abs/1303.3997">Li</a></li>
</ul>

<p>Alignment relies on the reference genome being <em>indexed</em> so that the sequencing reads can be located more efficiently. The genome index is a highly-accessible data structure, and Galaxy includes indices for many popular genomes. </p>

<h4>1.Align the example files</h4>

<p><img src="media/bowtie2-tool.png" alt="" title="" /></p>

<ul>
<li>Find the tool <em>NGS: Mapping</em> -> <em>Bowtie2</em>
<ul>
<li>alternatively, type <code>bowtie</code> in the search box</li>
</ul></li>
<li>In <em>Is this single-end or Paired-end?</em> Select <strong>Paired-end</strong></li>
<li>Set <em>FastQ file #1</em> and <em>FastQ file #2</em> to the two fastq files you uploaded in the previous step</li>
<li>Make sure the reference genome is set to <strong>Human (Homo sapiens)(b37):hg19</strong></li>
<li>Press <em>Execute</em></li>
<li>Wait!</li>
</ul>

<p>The result will be a <code>.bam</code> file that we will describe in the next section. This file is not human-readable, as it is compresed. But we can convert to a readable format for illustration purposes.</p>

<h4>2. View the alignments</h4>

<ol>
<li>Click on the eye of the resulting file to view the alignments.</li>
</ol>

<p><img src="media/bam-alignments.png" alt="" title="" /></p>

<h3>About the <code>bam</code> file format</h3>

<p>Unlike most of Bioinfomatics, a <em>single standard</em> file format has emerged for aligned reads. Moreoever, this file format is consistent regardless of whether you have DNA-seq, RNA-seq, ChIP-seq... data. </p>

<p>The <code>bam</code> file is a compressed, binary, version of a <code>sam</code> file.</p>

<h3>The <code>.sam</code> file</h3>

<ul>
<li><strong>S</strong>equence <strong>A</strong>lignment/<strong>M</strong>ap (sam) </li>
<li>The output from an aligner such as <code>bwa</code></li>
<li>Same format regardless of sequencing protocol (i.e. RNA-seq, ChIP-seq, DNA-seq etc)</li>
<li>May contain un-mapped reads</li>
<li>Potentially large size on disk; ~100s of Gb
<ul>
<li>Can be manipulated with standard unix tools; e.g. <em>cat</em>, <em>head</em>, <em>grep</em>, <em>more</em>, <em>less</em>....</li>
</ul></li>
<li>Official specification can be <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">obtained online</a>: -</li>
<li>We normally work on a compressed version called a <code>.bam</code> file. See later.</li>
<li><em>Header</em> lines starting with an <code>@</code> character, followed by tab-delimited lines
<ul>
<li>Header gives information about the alignment and references sequences used</li>
</ul></li>
</ul>

<p>The first part of the header lists the names (<code>SN</code>) of the sequences (chromosomes) used in alignment, their length (<code>LN</code>) and a <em>md5sum</em> "<a href="https://en.wikipedia.org/wiki/Md5sum">digital fingerprint</a>" of the <code>.fasta</code> file used for alignment (<code>M5</code>).</p>

<p>```</p>

<p>@HD VN:1.0 SO:coordinate
@SQ SN:chr10 LN:135534747
@SQ SN:chr11 LN:135006516
@SQ SN:chr11<em>gl000202</em>random LN:40103
@SQ SN:chr12 LN:133851895
@SQ SN:chr13 LN:115169878
@SQ SN:chr14 LN:107349540
@SQ SN:chr15 LN:102531392
@SQ SN:chr16 LN:90354753
.....
.....</p>

<p>```</p>

<p>We also have a section where we can record the processing steps used to derive the file
```
@PG ID:bowtie2 PN:bowtie2 VN:2.3.4.1 CL:"/jetstream/scratch0/main/conda/envs/mulled-v1-65d5efe4f1b69ab7166d1a5a5616adebe902133ea3e4c189d87d7de2e21ddc17/bin/bowtie2-align-s --wrapper basic-0 -p 10 -x /cvmfs/data.galaxyproject.org/byhand/hg19/hg19full/bowtie2<em>index/hg19full -1 input</em>f.fastq -2 input_r.fastq"
....
....</p>

<p>```</p>

<p>Next is a <em>tab-delimited</em> section that describes the alignment of each sequence in detail. </p>

<p>```
SRR081708.237649    163 1   10003   6   1S67M   =   10041   105 GACCCTGACCCTAACCCTGACCCTGACCCTAACCCTGACCCTGACCCTAACCCTGACCCTAACCCTAA    S=&lt;====&lt;&lt;>=><?=?=?>==@??;?>@@@=??@@????@??@?>?@@&lt;@>@'@=?=??=&lt;=>?>?=Q    ZA:Z:&lt;&amp;;0;0;;308;68M;68>&lt;@;0;0;;27;;>MD:Z:5A11A5A11A5A11A13 RG:Z:SRR081708  NM:i:6  OQ:Z:GEGFFFEGGGDGDGGGDGA?DCDD:GGGDGDCFGFDDFFFCCCBEBFDABDD-D:EEEE=D=DDDDC:</p>

<p>```</p>

<p><img src="media/sam-entry-explained.png" alt="" title="" /></p>

<p>Column | Official Name | Brief
------ | -------------- | -----------
1      | QNAME          | Sequence ID
2      | FLAG           | Sequence quality expressed as a bitwise flag
3      | RNAME          | Chromosome
4      | POS            | Start Position
5      | MAPQ           | Mapping Quality
6      | CIGAR          | Describes positions of matches, insertions, deletions w.r.t reference
7      | RNEXT          | Ref. name of mate / next read
8      | PNEXT          | Postion of mate / next read
9      | TLEN           | Observed Template length
10     | SEQ            | Sequence
11     | QUAL           | Base Qualities</p>

<p>There can also be all manner of optional tags as extra columns introduce by an aligner or downstream analysis tool. A common use is the <code>RG</code> tag which refers back to the read groups in the header.</p>

<h3>Dr Mark Dunning presents....Fun with flags!</h3>

<p>The <em>"flags"</em> in the sam file can represent useful QC information</p>

<ul>
<li>Read is unmapped</li>
<li>Read is paired / unpaired</li>
<li>Read failed QC</li>
<li>Read is a PCR duplicate (see later)</li>
</ul>

<p>The combination of any of these properties is used to derive a numeric value</p>

<p>For instance, a particular read has a flag of 163</p>

<p><img src="media/flag-highlight.png" alt="" title="" /></p>

<h3>Derivation</h3>

<p>There is a set of properties that a read can possess. If a particular property is observed, a corresponding power of 2 is added multiplied by 1. The final value is derived by summing all the powers of 2.</p>

<p><img src="https://galaxyproject.org/tutorials/ngs/sam_flag.png" alt="" title="" /></p>

<p>Flag Value | Meaning
---------- | --------------------------------
69 (= 1 + 4 + 64)   | The read is paired, is the first read in the pair, and is unmapped.
77 (= 1 + 4 + 8 + 64) | The read is paired, is the first read in the pair, both are unmapped.
83 (= 1 + 2 + 16 + 64) |    The read is paired, mapped in a proper pair, is the first read in the pair, and it is mapped to the reverse strand.
99 (= 1 + 2 + 32 + 64) |    The read is paired, mapped in a proper pair, is the first read in the pair, and its mate is mapped to the reverse strand.
133 (= 1 + 4 + 128) |   The read is paired, is the second read in the pair, and it is unmapped.
137 (= 1 + 8 + 128)  |  The read is paired, is the second read in the pair, and it is mapped while its mate is not.
141 (= 1 + 4 + 8 + 128) |   The read is paired, is the second read in the pair, but both are unmapped.
147 (= 1 + 2 + 16 + 128) |  The read is paired, mapped in a proper pair, is the second read in the pair, and mapped to the reverse strand.
163 (= 1 + 2 + 32 + 128) |  The read is paired, mapped in a proper pair, is the second read in the pair, and its mate is mapped to the reverse strand.</p>

<p>See also</p>

<ul>
<li>https://broadinstitute.github.io/picard/explain-flags.html</li>
</ul>

<h3>Have a CIGAR!</h3>

<p><img src="media/cigar-highlight.png" alt="" title="" /></p>

<p>The <strong><em>CIGAR</em></strong> (<strong>C</strong>ompact <strong>I</strong>diosyncratic <strong>G</strong>apped <strong>Alignment</strong> <strong>R</strong>eport) string is a way of encoding the match between a given sequence and the position it has been assigned in the genome. It is comprised by a series of letters and numbers to indicate how many consecutive bases have that mapping.</p>

<p>Code  | Description
------------- | -------------
M  | alignment match
I  | insertion
D  | deletion
N  | skipped
S  | soft-clipping
H  | hard-clipping</p>

<p>e.g.</p>

<ul>
<li><code>68M</code>
<ul>
<li>68 bases matching the reference</li>
</ul></li>
<li><code>1S67M</code>
<ul>
<li>1 soft-clipped read followed by 67 matches</li>
</ul></li>
<li><code>15M87N70M90N16M</code>
<ul>
<li>15 matches following by 87 bases skipped followed by 70 matches etc.</li>
</ul></li>
</ul>

<h4>3. Check the alignment stats</h4>

<p>We will now generate a few basic statistics about the alignment of our data</p>

<ol>
<li>Select the tool <em>NGS: SAMtools -> Flagstat</em> </li>
<li>In the <em>BAM File to Convert</em> box choose the bam file produced by bowtie2.</li>
</ol>

<p>The tool will also report how many <strong><em>PCR Duplicates</em></strong> have been found in the data. But as we haven't yet run any software to identify such reads, the flagstat output will show 0 reads.</p>

<h3>About Duplicates</h3>

<p>The preparation of a sequencing library requires <em>PCR</em> amplification of your starting material. This can lead to some DNA fragments being over-represented in your data. As our DNA fragments are formed in a random process, and relatively small compared to the number of bases to be sequenced from the genome (3Gb in humans), we tend to think the two DNA fragments that have identical starting and ending position are unlikely to have occured due to chance. Some software, such as <a href="http://broadinstitute.github.io/picard/">Picard</a> will identify such artefacts and <em>mark</em> them for attention by downstream methods. i.e. they are not completely discarded from the analysis.</p>

<p><img src="media/pcr_dups.png" alt="" title="" /></p>

<h4>4. Mark Duplicates with Picard</h4>

<ol>
<li>Use the tool <em>NGS: Picard -> MarkDuplicates</em></li>
<li>In <em>Select SAM/BAM dataset or dataset collection</em> choose the bam file produced by bowtie2.</li>
<li>What do you notice about the <em>flag</em> values for any reads that have the same <em>start</em> as another read? </li>
<li>Interpret the meaning of these flags using the online tool
<ul>
<li>https://broadinstitute.github.io/picard/explain-flags.html</li>
</ul></li>
</ol>

<p><strong>Warning</strong> the assumption about reads having the same start location being PCR duplicates falls down when we do sequencing for a very specific region of the genome. e.g. targeted sequencing from a panel of cancer genes. Running a tool to mark PCR duplicates on such data would recommend a high proportion of reads be ignored from further analysis.</p>

<h4>5. (Optional) Re-run the alignment statistics</h4>

<ol>
<li>Select the tool <em>NGS: SAMtools -> Flagstat</em> </li>
<li>In the <em>BAM File to Convert</em> box choose the bam file produced by the <em>mark duplicates</em> step</li>
</ol>
